README.txt
stompserver
    by Patrick Hurley, Lionel Bouton
    http://stompserver.rubyforge.org/

== DESCRIPTION:

Stomp messaging server with file/dbm/memory/activerecord based FIFO
queues, queue monitoring, and basic authentication.

== SYNOPSYS:

Handles basic message queue processing

== REQUIREMENTS:

* EventMachine

== FEATURES/PROBLEMS:

=== Several queue storage backends

Handles basic message queue processing using memory, file, or dbm
based queues. Messages are sent and consumed in FIFO order (unless a
client error happens, this should be corrected in the future). Topics
are memory-only storage.  You can select activerecord, file or dbm
storage and the queues will use that, but topics will only be stored
in memory.

memory queues are of course the fastest ones but shouldn't be used if
you want to ensure all messages are delivered.

dbm queues will use berkeleydb if available, otherwise dbm or gdbm
depending on the platform. sdbm does not work well with marshalled
data. Note that these queues have not been tested in this release.

For the file based storage, each frame is stored in a single file. The
first 8 bytes contains the header length, the next 8 bytes contains
the body length, then the headers are stored as a marshalled object
followed by the body stored as a string. This storage is currently
inefficient because queues are stored separately from messages, which
forces a double write for data safety reasons on each message stored.

The activerecord based storage expects to find a database.yml file in
the configuration directory. It should be the most robust backend, but
the slowest one. The database must have an ar_messages table which can
be created with the following code (you are responsible to do so):

  ActiveRecord::Schema.define do
    create_table 'ar_messages' do |t|
      t.column 'stomp_id', :string, :null => false
      t.column 'frame', :text, :null => false
    end
  end

You can read the frames with this model:

  class ArMessage < ActiveRecord::Base
    serialize :frame
  end

The ar_message implementation will certainly change in the future.

This is meant to be easily readable by a Rails application (which
could handle the ar_messages table creation with a migration).

=== Limitations

Stompserver not support any server to server messaging (although you could
write a client to do this).

=== Monitoring

Queues can be monitored via the monitor queue (this will probably not
be supported this way in the future to avoid polluting the queue
namespace). If you subscribe to /queue/monitor, you will receive a
status message every 5 seconds that displays each queue, it's size,
frames enqueued, and frames dequeued. Stats are sent in the same
format of stomp headers, so they are easy to parse. Following is an
example of a status message containing stats for 2 queues:

Queue: /queue/client2
size: 0
dequeued: 400
enqueued: 400

Queue: /queue/test
size: 50
dequeued: 250
enqueued: 300

=== Access control

Basic client authorization is also supported.  If the -a flag is
passed to stompserver on startup, and a .passwd file exists in the run
directory, then clients will be required to provide a valid login and
passcode.  See passwd.example for the password file format.

=== Misc

Whenever you stop the server, any queues with no messages will be
removed, and the stats for that queue will be reset.  If the queue has
any messages remaining then the stats will be saved and available on
the next restart.

== INSTALL:

* gem install stompserver

stompserver will create a log, etc, and storage directory on startup
in your current working directory, the value passed to as
--working_dir parameter, or if using a config file it will
use what you specified for working_dir.  The configuration file is a
yaml file and can be specified on the command line with -C
<configfile>. A sample is provided in config/stompserver.conf.

Command line options will override options set in the yaml config
file.

To use the memory queue run as follows:
  stompserver -p 61613 -b 0.0.0.0

To use the file or dbm queue storage, use the -q switch and specificy
either file or dbm.  The file and dbm queues also take a storage
directory specified with -s.  .stompserver is the default directory if
-s is not used.
  stompserver -p 61613 -b 0.0.0.0 -q file -s .stompfile
Or
  stompserver -p 61613 -b 0.0.0.0 -q dbm -s .stompbdb

To specify where the queue is stored on disk, use the -s flag followed
by a storage directory.  To enable client authorization use -a, for
debugging use -d.
  stompserver -p 61613 -b 0.0.0.0 -q file -s .stompserver -a -d

You cannot use the same storage directory for a file and dbm queue,
they must be kept separate.

To use the activerecord queue storage use -q activerecord:
  stompserver -p 61613 -b 0.0.0.0 -q activerecord
It will try to read the etc/database.yml file in the working
directory. Here's an example of a database.yml for a PostgreSQL
database named stompserver on the 'dbserver' host usable by
PostgreSQL's user 'foo' with password 'bar'(see ActiveRecord's
documentation for the parameters needed by your database):

  adapter: postgresql
  database: stompserver
  username: foo
  password: bar
  host: dbserver

== LICENSE:

(The MIT License)

Copyright (c) 2006 Patrick Hurley
Copyright (c) 2007 Lionel Bouton
Copyright (c) 2010 Tyler Green

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 0.9.9.1 / 30 Nov 2010

* Added timeouts for messages
* Storage backends are now required to implement a 'destinations' method

== 0.9.8 / 16 Aug 2007

* Several storage backends instead of Madeleine (allow tradeoffs between
  robustness with ActiveRecord and speed with memory with intermediary
  solutions based on file or dbm storages).
* Better load distribution between clients acknowledging messages after
  processing them.
* Monitoring support through a dedicated queue.

== 0.9.5 / 18 Oct 2006

* Removed eventmachine dependency in the gem
  you still need eventmachine, but you can choose between
  eventmachine and eventmachine-win32, not sure how to put 
  that in the gem.
  
  Also made another minor fix to the stompserver binary.

== 0.9.4 / 17 Oct 2006

* Cleanup (wow I was tired last night)
  * Tested under Linux, added require 'rubygems' 
  * Fixed stompserver binary - it now works and takes a few simple command args
  * Fixed up minor (tesly) test syntax issues

== 0.9.3 / 16 Oct 2006

* Last one tonight I promise
  * Fixed incorrect gem requirement version on eventmachine
  * Fixed stompserver cmd added by gem

== 0.9.2 / 16 Oct 2006

* More Queue issues resolved
  * Added more unit tests
  * All tests pass
  * Queue processing done in one large batch, this may change

== 0.9.1 / 16 Oct 2006

* Stepping along
  * Fixed release issues
  * Fixed queue logic error (failed to detect offlines)

== 0.9.0 / 16 Oct 2006

* Initialial Beta Release
  * Seems to work
  * Passes numerous test cases
  * Journals using madeleine
  * Needs documentaion
  * Needs to command line processing
  * Needs service/daemon options4



lib/stomp_server.rb
require 'rubygems'
require 'eventmachine'
require 'stomp_server/stomp_frame'
require 'stomp_server/stomp_id'
require 'stomp_server/stomp_auth'
require 'stomp_server/topic_manager'
require 'stomp_server/queue_manager'
require 'stomp_server/queue'
require 'stomp_server/queue/memory_queue'
require 'stomp_server/queue/file_queue'
require 'stomp_server/queue/dbm_queue'
require 'stomp_server/protocols/stomp'

module StompServer
  VERSION = '0.9.9'

  class Configurator
    attr_accessor :opts

    def initialize
      @opts = nil
      @defaults = {
        :port => 61613,
        :host => "127.0.0.1",
        :debug => false,
        :queue => 'memory',
        :auth => false,
        :working_dir => Dir.getwd,
        :storage => ".stompserver",
        :logdir => 'log',
        :monitor_sleep_time => 5,
        :configfile => 'stompserver.conf',
        :logfile => 'stompserver.log',
        :pidfile => 'stompserver.pid',
        :checkpoint => 0,
        :timeout => 60,
        :timeout_check_freq => 30
      }
      @opts = getopts
      if opts[:debug]
        $DEBUG=true
      end

    end

    def getopts
      copts = OptionParser.new
      copts.on("-C", "--config=CONFIGFILE", String, "Configuration File (default: stompserver.conf)") {|c| @defaults[:configfile] = c}
      copts.on("-p", "--port=PORT", Integer, "Change the port (default: 61613)") {|p| @defaults[:port] = p}
      copts.on("-m", "--monitor-interval=TIME", Float, "Change the sleep time between monitor updates") {|p| @defaults[:monitor_sleep_time] = p}
      copts.on("-b", "--host=ADDR", String, "Change the host (default: localhost)") {|a| @defaults[:host] = a}
      copts.on("-q", "--queuetype=QUEUETYPE", String, "Queue type (memory|dbm|activerecord|file) (default: memory)") {|q| @defaults[:queue] = q}
      copts.on("-w", "--working_dir=DIR", String, "Change the working directory (default: current directory)") {|s| @defaults[:working_dir] = s}
      copts.on("-s", "--storage=DIR", String, "Change the storage directory (default: .stompserver, relative to working_dir)") {|s| @defaults[:storage] = s}
      copts.on("-d", "--debug", String, "Turn on debug messages") {|d| @defaults[:debug] = true}
      copts.on("-a", "--auth", String, "Require client authorization") {|a| @defaults[:auth] = true}
      copts.on("-c", "--checkpoint=SECONDS", Integer, "Time between checkpointing the queues in seconds (default: 0)") {|c| @defaults[:checkpoint] = c}
      copts.on("-t", "--timeout=SECONDS", Integer, 'Number of seconds to keep an unsent message in a queue') {|t| @defaults[:timeout] = t}
      copts.on("-f", "--freq=SECONDS", Integer, 'Interval between checking queues for timeout messages') {|f| @defaults[:timeout_check_freq] = f }
      copts.on("-h", "--help", "Show this message") do
        puts copts
        exit
      end
      puts copts.parse(ARGV)

      if File.exists?(@defaults[:configfile])
        opts = @defaults.merge(YAML.load_file(@defaults[:configfile]))
      else
        opts = @defaults
      end

      opts[:etcdir] = File.join(opts[:working_dir],'etc')
      opts[:storage] = File.join(opts[:working_dir],opts[:storage])
      opts[:logdir] = File.join(opts[:working_dir],opts[:logdir])
      opts[:logfile] = File.join(opts[:logdir],opts[:logfile])
      opts[:pidfile] = File.join(opts[:logdir],opts[:pidfile])
      if opts[:auth]
        opts[:passwd] = File.join(opts[:etcdir],'.passwd')
      end

      return opts
    end
  end


  class Run
    attr_accessor :queue_manager, :auth_required, :stompauth, :topic_manager

    def initialize(opts)
      @opts = opts
      @queue_manager = nil
      @auth_required = nil
      @stompauth = nil
      @topic_manager = nil
    end

    def stop(pidfile)
      @queue_manager.stop
      puts "Stompserver shutting down" if $DEBUG
      EventMachine::stop_event_loop
      File.delete(pidfile)
    end

    def start
      begin
        if @opts[:group]
          puts "Changing group to #{@opts[:group]}."
          Process::GID.change_privilege(Etc.getgrnam(@opts[:group]).gid)
        end

        if @opts[:user]
          puts "Changing user to #{@opts[:user]}."
          Process::UID.change_privilege(Etc.getpwnam(@opts[:user]).uid)
        end
      rescue Errno::EPERM
        puts "FAILED to change user:group #{@opts[:user]}:#{@opts[:group]}: #$!"
        exit 1
      end
      
     [:working_dir,
      :logdir,
      :etcdir
     ].each { |dir|  Dir.mkdir(@opts[dir]) unless File.directory?(@opts[dir]) }

      if @opts[:daemon]
        Daemonize.daemonize(log_file=@opts[:logfile])
        # change back to the original starting directory
        Dir.chdir(@opts[:working_dir])
      end

      # Write pidfile
      open(@opts[:pidfile],"w") {|f| f.write(Process.pid) }
      
      qstore = case @opts[:queue]
               when 'dbm'
                 StompServer::DBMQueue.new(@opts[:storage])
               when 'file'
                 StompServer::FileQueue.new(@opts[:storage])
               when 'activerecord'
                 require 'stomp_server/queue/activerecord_queue'
                 StompServer::ActiveRecordQueue.new(@opts[:etcdir], @opts[:storage])
               else
                 StompServer::MemoryQueue.new
               end

      qstore.checkpoint_interval = @opts[:checkpoint]
      puts "Checkpoing interval is #{qstore.checkpoint_interval}" if $DEBUG
      @topic_manager = StompServer::TopicManager.new
      @queue_manager = StompServer::QueueManager.new(qstore, 
                                                     :timeout => @opts[:timeout],
                                                     :timeout_check_freq => @opts[:timeout_check_freq],
                                                     :monitor_sleep_time => @opts[:monitor_sleep_time])
      @auth_required = @opts[:auth]

      if @auth_required
        @stompauth = StompServer::StompAuth.new(@opts[:passwd])
      end

      trap("INT") { puts "INT signal received.";stop(@opts[:pidfile]) }
    end
  end
end

